File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        mn=-1e18
        mx=1e18
        ans=True
        
        def find(root,mn,mx):
            nonlocal ans
            if root==None:
                return
            
            find(root.left,mn,root.val)
            
            if root.val<=mn or root.val>=mx:
                ans=False
                return
            
            find(root.right,root.val,mx)
        
        find(root,mn,mx)
        
        return ans
        

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#define lln long long int

class Solution {
public:
    bool first=false;
    int prev;
    bool isValidBST(TreeNode* root) {
       
        bool ans=true;
        
        lln l=-1e15;
        
        lln r=1e15;
        
        find(root,ans,l,r);
        
        return ans;
    }
    

    void find(TreeNode* root,bool &ans,lln mn,lln mx){
        if(!root)
            return;
        
     find(root->left,ans,mn,root->val);
     
     if(root->val<=mn || root->val >=mx){
         ans=false;
         return;
     }
        
     find(root->right,ans,root->val,mx);
        
    }
};